src/rules/mapreduce.jl

```diff
@@ -0,0 +1,30 @@
+#####
+##### `map`
+#####
+
+function rrule(::typeof(map), f, xs...)
+    T = Tuple{Core.Typeof(f), map(eltype, xs)...}
+    if hasmethod(rrule, T)
```

Comment:
You should actually call `rrule`/`frule` and check the result, instead of trying to use reflection to drive closure creation. In general, we should try to discourage unnecessary reflection; that was one of the old DiffRules problems ChainRules set out to fix.

Reply:
The reason I did it this way was to avoid having the check inside of the `map`ped function, so that we don't do the check at every iteration. But that's probably preferable to this mess anyway.

Reply:
> was to avoid having the check inside of the mapped function

The no-rule case (returning `nothing`) is very compiler-friendly; an `isa Nothing` check should get compiled away. But even if it wasn't, AFAICT, the reflection approach seems incorrect for arguments with non-concrete eltypes (rules work even if they are dynamically dispatched to)

Reply:
I dumped all of this crazy stuff in favor of a simple check within the mapped function inside the rule.